languages/emf-metamodel/src/main/java/de/jplag/emf/util/GenericEmfTreeView.java

```diff
@@ -0,0 +1,153 @@
+package de.jplag.emf.util;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.StringJoiner;
+
+import org.eclipse.emf.ecore.ENamedElement;
+import org.eclipse.emf.ecore.EObject;
+import org.eclipse.emf.ecore.EStructuralFeature;
+import org.eclipse.emf.ecore.resource.Resource;
+
+import de.jplag.TokenTrace;
+import de.jplag.emf.MetamodelToken;
+
+/**
+ * Very basic tree view representation of an EMF metamodel or model.
+ */
+public class GenericEmfTreeView extends AbstractModelView {
+    private final List<String> lines;
+    private final Map<EObject, TokenTrace> objectToLine;
+    private final ModelingElementIdentifierManager identifierManager;
+
+    /**
+     * Creates a tree view for a metamodel.
+     * @param file is the path to the metamodel.
+     */
+    public GenericEmfTreeView(File file, Resource modelResource) {
+        super(file);
+        lines = new ArrayList<>();
+        objectToLine = new HashMap<>();
+        identifierManager = new ModelingElementIdentifierManager();
+        TreeViewBuilder visitor = new TreeViewBuilder();
+        modelResource.getContents().forEach(visitor::visit);
+    }
+
+    /**
+     * Adds a token to the view, thus adding the index information to the token. Returns a new token enriched with the index
+     * metadata.
+     * @param token is the token to add.
+     */
+    @Override
+    public MetamodelToken convertToMetadataEnrichedToken(MetamodelToken token) {
+        Optional<EObject> optionalEObject = token.getEObject();
+        if (optionalEObject.isPresent()) {
+            EObject object = optionalEObject.get();
+            TokenTrace trace = objectToLine.get(object);
+            return new MetamodelToken(token.getType(), token.getFile(), trace, optionalEObject);
+        }
+        return new MetamodelToken(token.getType(), token.getFile());
+    }
+
+    private final class TreeViewBuilder extends AbstractMetamodelVisitor {
+        private static final String IDENTIFIER_PREFIX = " #";
```

Comment:
Is there no easier way of doing this?

Reply:
What do you mean?

Reply:
It is harder to read the code when you have to scroll back to the top to see what is added and what is the result